lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java

```diff
@@ -1009,6 +1005,32 @@ public void logCurrentState() {
     }
 
     @Override
+    public void logCurrentHeldLocks(String descriptorRegex, boolean versionIdNecessary) {
+        StringBuilder logString = getGeneralLockStats();
+        Pattern pattern = Pattern.compile(descriptorRegex);
+        Set<ExpiringToken> heldLocksTokens = ImmutableSet.<ExpiringToken>builder()
+                .addAll(heldLocksTokenMap.keySet())
+                .addAll(heldLocksGrantMap.keySet())
+                .build();
+
+        for (ExpiringToken heldLocksToken : heldLocksTokens) {
```

Comment:
`heldLocksTokens.stream().filter(!versionIdNecessary || heldLocksToken.getVersionId() != null).map(ExpiringToken::getLockDescriptors).forEach(lambda)`

Where `lambda` is lines 1019-1024 extracted into a method.
`logString` can be a field in the little `LockStatsGenerator` class.